lib/elixir/lib/kernel.ex

```diff
@@ -1502,7 +1509,27 @@ defmodule Kernel do
 
   """
   defmacro left and right do
-    quote(do: :erlang.andalso(unquote(left), unquote(right)))
+    case __CALLER__.context do
+      nil -> build_boolean_check(:and, left, right, false)
+      _ -> quote(do: :erlang.andalso(unquote(left), unquote(right)))
+    end
+  end
+
+  defp build_boolean_check(operator, check, true_clause, false_clause) do
+    optimize_boolean(
+      quote do
+        case unquote(check) do
+          true ->
```

Comment:
I think this case it may be worth inlining everything?

```
case unquote(check) do
  true -> unquote(true_clause)
  false -> unquote(false_clause)
end
```

?